Math/src/main/java/org/tribuo/math/distributions/MultivariateNormalDistribution.java

```diff
@@ -91,62 +99,278 @@ public MultivariateNormalDistribution(DenseVector means, DenseMatrix covariance,
+                        this.samplingCovariance = eigenvectors.matrixMultiply(diagonal).matrixMultiply(eigenvectors,false,true);
+                        this.determinant = factorization.get().determinant();
+                        this.precision = factorization.get().inverse();
+                    } else {
+                        throw new IllegalArgumentException("Covariance matrix is not positive definite.");
+                    }
+                } else {
+                    Optional<DenseMatrix.CholeskyFactorization> factorization = this.covarianceMatrix.choleskyFactorization();
+                    if (factorization.isPresent()) {
+                        this.samplingCovariance = factorization.get().lMatrix();
+                        this.determinant = factorization.get().determinant();
+                        this.precision = factorization.get().inverse();
+                    } else {
+                        throw new IllegalArgumentException("Covariance matrix is not positive definite.");
+                    }
+                }
+                // set unused variables.
+                this.covarianceVector = null;
+                this.variance = Double.NaN;
+            }
+            case DIAGONAL -> {
+                if (!(covariance instanceof DenseVector)) {
+                    throw new IllegalArgumentException("Covariance must be a vector for diagonal covariance, found " + covariance.getClass());
+                }
+                this.covarianceVector = (DenseVector) covariance.copy();
+                if (this.covarianceVector.size() != this.means.size()) {
+                    throw new IllegalArgumentException("Covariance must be a vector and the same dimension as the mean vector. Mean vector size = " + means.size() + ", covariance size = " + this.covarianceVector.size());
+                }
+
+                double tmp = 1;
+                for (int i = 0; i < this.covarianceVector.size(); i++) {
+                    tmp *= this.covarianceVector.get(i);
+                }
+                this.determinant = tmp;
+                this.precision = this.covarianceVector.copy();
+                this.precision.foreachInPlace(a -> 1.0/a);
+
+                // set unused variables.
+                this.covarianceMatrix = null;
+                this.samplingCovariance = null;
+                this.variance = Double.NaN;
             }
-        } else {
-            Optional<DenseMatrix.CholeskyFactorization> factorization = this.covariance.choleskyFactorization();
-            if (factorization.isPresent()) {
-                this.samplingCovariance = factorization.get().lMatrix();
-            } else {
-                throw new IllegalArgumentException("Covariance matrix is not positive definite.");
+            case SPHERICAL -> {
+                if (covariance instanceof DenseVector vec) {
+                    if ((vec.size() != 1) && (vec.size() != means.size())) {
```

Comment:
I'm confused by the second clause here - shouldn't the first one be the only one needed?

Reply:
It allows the covariance to be a vector of length means with all the same value. Though I'm not checking that they are the same value here.